SQS-379 | unit tests for processOrderBookActiveOrders

deividaspetraitis opened the pull request:
Implements tests for processOrderBookActiveOrders orderbook usecase method. As a side effect introduces some helper methods and refactors some test cases.

deividaspetraitis committed:
SQS-379 | processOrderBookActiveOrders

Implements tests for processOrderBookActiveOrders orderbook usecase method.
As a side effect introduces some helper methods and refactors some test
cases.

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -401,42 +457,144 @@ func (s *OrderbookUsecaseTestSuite) TestGetActiveOrders() {
-			tickValues := orderbookdomain.TickValues{
-				EffectiveTotalAmountSwapped: effectiveTotalAmountSwapped,
+func (s *OrderbookUsecaseTestSuite) TestProcessOrderBookActiveOrders() {
+	withGetMetadataByChainDenom := func(tokensusecase *mocks.TokensUsecaseMock, errIfNotDenom string) *mocks.TokensUsecaseMock {
```
Some comments here would help to grok what's going on 🙏

deividaspetraitis replied:
Added comments, please check [here](https://github.com/osmosis-labs/sqs/pull/499/files#diff-5fa55bd6410efa835405589edaffbb41ba95f810344564647d2d711e8bc15b67R191)! :pray:

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -35,6 +35,101 @@ func TestOrderbookUsecaseTestSuite(t *testing.T) {
 	suite.Run(t, new(OrderbookUsecaseTestSuite))
 }
 
+// Creates a new osmomath.Dec from a string
+// Panics if the string is invalid
+func (s *OrderbookUsecaseTestSuite) newDecFromStr(str string) osmomath.Dec {
+	s.T().Helper()
+	dec, err := osmomath.NewDecFromStr(str)
+	s.Require().NoError(err)
+	return dec
+}
+
```
Consider directly using and removing the helper:
```go
osmomath.MustNewDecFromStr(...)
```

deividaspetraitis replied:
Updated accordingly :+1:

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -35,6 +35,101 @@ func TestOrderbookUsecaseTestSuite(t *testing.T) {
+	return dec
+}
+
+func (s *OrderbookUsecaseTestSuite) newOrder() orderbookdomain.Order {
+	s.T().Helper()
+
+	return orderbookdomain.Order{
+		TickId:         1,
+		OrderId:        1,
+		OrderDirection: "bid",
+		Owner:          "owner1",
+		Quantity:       "1000",
+		PlacedQuantity: "1500",
+		Etas:           "500",
+		ClaimBounty:    "10",
+		PlacedAt:       "1634764800000",
+	}
+}
```
Why do we need to define it as a method/helper?

Could we for simplicity define the variable at the top of the file instead?

i.e.
```go
var defaultOrder = orderbookdomain.Order{
  TickId:         1,
  OrderId:        1,
  OrderDirection: "bid",
  Owner:          "owner1",
  Quantity:       "1000",
  PlacedQuantity: "1500",
  Etas:           "500",
  ClaimBounty:    "10",
  PlacedAt:       "1634764800000",
}
```

deividaspetraitis replied:
Just did that!

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -35,6 +35,101 @@ func TestOrderbookUsecaseTestSuite(t *testing.T) {
+	}
+}
+
+func (s *OrderbookUsecaseTestSuite) newLimitOrder() orderbookdomain.LimitOrder {
```
Ditto about helper

deividaspetraitis replied:
Moved to [OrderbookTestHelper](https://github.com/osmosis-labs/sqs/pull/499/files#diff-5fa55bd6410efa835405589edaffbb41ba95f810344564647d2d711e8bc15b67R123), please let me does it make sense.

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -35,6 +35,101 @@ func TestOrderbookUsecaseTestSuite(t *testing.T) {
+	}
+}
+
+func (s *OrderbookUsecaseTestSuite) newTick(effectiveTotalAmountSwapped string, unrealizedCancels int64, direction string) orderbookdomain.OrderbookTick {
```
This one makes sense to be a helper IMO - nice work!

However, could you please add some comments to make this more grokable?

deividaspetraitis replied:
Added some docs/comments :+1:

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -401,42 +457,144 @@ func (s *OrderbookUsecaseTestSuite) TestGetActiveOrders() {
 	}
+}
 
+func (s *OrderbookUsecaseTestSuite) TestCreateFormattedLimitOrder() {
```
Nice clean up for this test.

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -401,42 +457,144 @@ func (s *OrderbookUsecaseTestSuite) TestGetActiveOrders() {
+				PoolID:          1,
+				Quote:           order.QuoteAsset.Symbol,
+				Base:            order.BaseAsset.Symbol,
+			}, "ownerAddress")
```
Similar to comments in other PRs, I think we should have one test case where we vary the `ownderAddress`

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -401,42 +457,144 @@ func (s *OrderbookUsecaseTestSuite) TestGetActiveOrders() {
+
+			// Assert the results
+			if tc.expectedError != nil {
+				s.Assert().Error(err)
+				if errors.Is(err, tc.expectedError) {
+					s.Assert().ErrorIs(err, tc.expectedError)
+				} else {
+					s.Assert().ErrorAs(err, tc.expectedError)
+				}
```
I've seen this logic in a previous PR review - I think it makes sense to move it into a helper defined on `routertesting` to reduce code duplication.

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -401,42 +457,144 @@ func (s *OrderbookUsecaseTestSuite) TestGetActiveOrders() {
+			order := s.newLimitOrder()
+
+			// Call the method under test
+			orders, isBestEffort, err := usecase.ProcessOrderBookActiveOrders(context.Background(), domain.CanonicalOrderBooksResult{
+				ContractAddress: order.OrderbookAddress,
+				PoolID:          1,
+				Quote:           order.QuoteAsset.Symbol,
+				Base:            order.BaseAsset.Symbol,
```
Ditto for at least one case where test parameters are diverging from default to make sure that there is no logic, the success of which is coupled to a specific value of a parameter

deividaspetraitis replied:
I have [moved](https://github.com/osmosis-labs/sqs/pull/499/files#diff-5fa55bd6410efa835405589edaffbb41ba95f810344564647d2d711e8bc15b67R654) `order` as a test case parameter.  :+1:

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -401,42 +457,144 @@ func (s *OrderbookUsecaseTestSuite) TestGetActiveOrders() {
+			expectedIsBestEffort: true,
+		},
+		{
+			name: "successful processing of active orders",
```
Would love to see 2 extra test cases:
- 2 successful orders processed
- 2 processed but one is best-effort

deividaspetraitis replied:
For 2 orders created this [one](https://github.com/osmosis-labs/sqs/pull/499/files#diff-5fa55bd6410efa835405589edaffbb41ba95f810344564647d2d711e8bc15b67R749), and updated best [effort](https://github.com/osmosis-labs/sqs/pull/499/files#diff-5fa55bd6410efa835405589edaffbb41ba95f810344564647d2d711e8bc15b67R705).

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/SQS-378' into SQS-379

deividaspetraitis committed:
SQS-379 | processOrderBookActiveOrders

Implements requested changes

deividaspetraitis committed:
SQS-379 | processOrderBookActiveOrders

Adds some docs.

deividaspetraitis committed:
SQS-379 | processOrderBookActiveOrders

Cleanup.

deividaspetraitis committed:
SQS-379 | processOrderBookActiveOrders

Move OrderbookTestHelper and friends to orderbooktesting package.

p0mvn reviewed (approved):
Good work on test coverage and keeping the test readable!

deividaspetraitis committed:
SQS-346 | handler tests, swagger API (#501)

* SQS-346 | handler tests, swagger API

Implements unit tests for the orderbook active orders handler and
adds swagger API for the endpoint
